File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/EmailAddressValidator.vb
Public Class EmailAddressValidator
    Inherits DataAnnotationsModelValidator(Of EmailAddressAttribute)

    Public Sub New(ByVal metadata As ModelMetadata,
                   ByVal context As ControllerContext, ByVal attribute As EmailAddressAttribute)
        MyBase.New(metadata, context, attribute)
    End Sub

    Public Overrides Function GetClientValidationRules() As IEnumerable(Of ModelClientValidationRule)
        Dim rule = New ModelClientValidationRegexRule(
                   Me.ErrorMessage, EmailAddressAttribute.EmailRegex)

        Return New ModelClientValidationRule() {rule}
    End Function
End Class

File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/EmailAddressValidator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace CoolMvcBlog.Utils
{
    public class EmailAddressValidator : DataAnnotationsModelValidator<EmailAddressAttribute>
    {
        public EmailAddressValidator(
            ModelMetadata metadata, ControllerContext context, EmailAddressAttribute attribute)
            : base(metadata, context, attribute)
        {
        }

        public override IEnumerable<ModelClientValidationRule> GetClientValidationRules()
        {
            var rule = new ModelClientValidationRegexRule(this.ErrorMessage, EmailAddressAttribute.EmailRegex);

            yield return rule;
        }
    }
}